benchmark/worker/benchmark_client.go

```diff
@@ -0,0 +1,330 @@
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package main
+
+import (
+	"math"
+	"runtime"
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/benchmark"
+	testpb "google.golang.org/grpc/benchmark/grpc_testing"
+	"google.golang.org/grpc/benchmark/stats"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/grpclog"
+)
+
+var (
+	caFile = "benchmark/server/testdata/ca.pem"
+)
+
+type benchmarkClient struct {
+	stop          chan bool
+	mu            sync.RWMutex
+	lastResetTime time.Time
+	histogram     *stats.Histogram
+}
+
+func startBenchmarkClientWithSetup(setup *testpb.ClientConfig) (*benchmarkClient, error) {
+	var opts []grpc.DialOption
+
+	// Some setup options are ignored:
+	// - client type:
+	//     will always create sync client
+	// - async client threads.
+	// - core list
+	grpclog.Printf(" * client type: %v (ignored, always creates sync client)", setup.ClientType)
+	switch setup.ClientType {
+	case testpb.ClientType_SYNC_CLIENT:
+	case testpb.ClientType_ASYNC_CLIENT:
+	default:
+		return nil, grpc.Errorf(codes.InvalidArgument, "unknow client type: %v", setup.ClientType)
+	}
+	grpclog.Printf(" * async client threads: %v (ignored)", setup.AsyncClientThreads)
+	grpclog.Printf(" * core list: %v (ignored)", setup.CoreList)
+
+	grpclog.Printf(" - security params: %v", setup.SecurityParams)
+	if setup.SecurityParams != nil {
+		creds, err := credentials.NewClientTLSFromFile(Abs(caFile), setup.SecurityParams.ServerHostOverride)
+		if err != nil {
+			return nil, grpc.Errorf(codes.InvalidArgument, "failed to create TLS credentials %v", err)
+		}
+		opts = append(opts, grpc.WithTransportCredentials(creds))
+	} else {
+		opts = append(opts, grpc.WithInsecure())
+	}
+
+	grpclog.Printf(" - core limit: %v", setup.CoreLimit)
+	// Use one cpu core by default
+	numOfCores := 1
+	if setup.CoreLimit > 0 {
+		numOfCores = int(setup.CoreLimit)
+	}
+	runtime.GOMAXPROCS(numOfCores)
```

Comment:
replace the above with:

if config.CoreLimit > 1 {
  runtime.GoMAXPROCS(config.CoreLimit)
}